lib/evaluate/characteristics.h

```diff
@@ -26,38 +26,88 @@
+
+// Absent components are deferred or assumed.
+using Shape = std::vector<std::optional<Expr<SubscriptInteger>>>;
+
+class TypeAndShape {
+public:
+  explicit TypeAndShape(DynamicType t) : type_{t} {}
+  DEFAULT_CONSTRUCTORS_AND_ASSIGNMENTS(TypeAndShape)
+
+  DynamicType type() const { return type_; }
+  const Shape &shape() const { return shape_; }
+  bool IsAssumedRank() const { return isAssumedRank_; }
+
+  bool operator==(const TypeAndShape &) const;
+
+  static std::optional<TypeAndShape> Characterize(const semantics::Symbol &);
+  static std::optional<TypeAndShape> Characterize(const semantics::Symbol *);
+  static std::optional<TypeAndShape> Characterize(
+      const semantics::ObjectEntityDetails &);
+  static std::optional<TypeAndShape> Characterize(
+      const semantics::ProcEntityDetails &);
+  static std::optional<TypeAndShape> Characterize(
+      const semantics::ProcInterface &);
+  static std::optional<TypeAndShape> Characterize(
+      const semantics::DeclTypeSpec &);
+  static std::optional<TypeAndShape> Characterize(
+      const semantics::DeclTypeSpec *);
+
+  std::ostream &Dump(std::ostream &) const;
+
+private:
+  void AcquireShape(const semantics::ObjectEntityDetails &);
+
+protected:
+  DynamicType type_;
+  Shape shape_;
+  bool isAssumedRank_{false};
+};
 
 // 15.3.2.2
-struct DummyDataObject {
-  ENUM_CLASS(Attr, AssumedRank, Optional, Allocatable, Asynchronous, Contiguous,
-      Value, Volatile, Polymorphic, Pointer, Target)
+struct DummyDataObject : public TypeAndShape {
+  ENUM_CLASS(Attr, Optional, Allocatable, Asynchronous, Contiguous, Value,
+      Volatile, Pointer, Target)
   DEFAULT_CONSTRUCTORS_AND_ASSIGNMENTS(DummyDataObject)
-  DynamicType type;
-  std::vector<std::optional<Expr<SubscriptInteger>>> shape;
+  explicit DummyDataObject(const TypeAndShape &t) : TypeAndShape{t} {}
+  explicit DummyDataObject(TypeAndShape &&t) : TypeAndShape{std::move(t)} {}
+  explicit DummyDataObject(DynamicType t) : TypeAndShape{t} {}
+  bool operator==(const DummyDataObject &) const;
+  static std::optional<DummyDataObject> Characterize(const semantics::Symbol &);
+  std::ostream &Dump(std::ostream &) const;
   std::vector<Expr<SubscriptInteger>> coshape;
   common::Intent intent{common::Intent::Default};
   common::EnumSet<Attr, 32> attrs;
-  bool operator==(const DummyDataObject &) const;
-  std::ostream &Dump(std::ostream &) const;
 };
 
 // 15.3.2.3
 struct DummyProcedure {
   ENUM_CLASS(Attr, Pointer, Optional)
   DECLARE_CONSTRUCTORS_AND_ASSIGNMENTS(DummyProcedure)
-  common::CopyableIndirection<Procedure> explicitProcedure;
-  common::EnumSet<Attr, 32> attrs;
+  explicit DummyProcedure(Procedure &&);
   bool operator==(const DummyProcedure &) const;
+  static std::optional<DummyProcedure> Characterize(
+      const semantics::Symbol &, const IntrinsicProcTable &);
   std::ostream &Dump(std::ostream &) const;
+  common::CopyableIndirection<Procedure> procedure;
+  common::EnumSet<Attr, 32> attrs;
```

Comment:
Why not symbols::Attr_enumSize instead of 32?

Reply:
Sure, will do here and below.

Reply:
I missed that Attr is a nested enum, so it should just be Attr_enumSize.